experimental/bot/logger/telemetrylogger/telemetry_logger.go

```diff
@@ -0,0 +1,74 @@
+package telemetrylogger
+
+import (
+	"fmt"
+
+	"github.com/mattermost/mattermost-plugin-api/experimental/bot/logger"
+	"github.com/mattermost/mattermost-plugin-api/experimental/telemetry"
+)
+
+type telemetryLogger struct {
+	logger.Logger
+	logLevel logger.LogLevel
+	tracker  telemetry.Tracker
+}
+
+/*
+New promotes the provided logger into a telemetry logger, storing all events below the logLevel
+through the tracker.
+
+- l Logger: A logger to promote.
+
+- logLevel: The highest type of message to be stored in telemetry.
+
+- tracker: The telemetry tracker to store the messages.
+*/
+func New(l logger.Logger, logLevel logger.LogLevel, tracker telemetry.Tracker) logger.Logger {
```

Comment:
It's idiomatic in go to return actual implementations instead of interfaces. What is the reasoning behind not doing this here? (`telemetryLogger` should be exported then)

Reply:
One upside of returning the interface is that we are sure on compile time that we are conformant with the interface. Nevertheless, there are other ways to achieve the same.

The main reason is personal preference. I prefer to keep things as "standard" as possible so they can be generalized. If we start returning the struct, we might end up declaring directly the struct, or modifying things that we understand invariant on the struct and making it more prone to errors. If we just interact with the interface, the usage is more stiff, but it is harder to "make mistakes".

Also, I would not like other packages working over "telemetryLogger" objects, since will make things way harder to mock and generalize.

Nevertheless, as I said, it is a personal preference. I am open to discussion or changes.

Reply:
I've seen strong opinions expressed on the subject. I do not have a personal strong preference, but lean towards the public `NewXXX` methods returning (public) interfaces. As practical, `newXXX` returning concrete types can be added for intra-package use.

Reply:
I don't have a strong option either. Keeping things as generalized as possible is a good point :+1: